Vectorized Convert Rotations node

kdsin23 opened the pull request:
Move the fuctions to the right place

sorry for this ,i feel patch-1 is a lot messy by me, so i pull a new request.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pxd:
```diff
@@ -6,4 +6,9 @@ from . vector cimport Vector3
-cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, float cosAngle)
\ No newline at end of file
+cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, float cosAngle)
+cdef euler3ToQuaternion(Quaternion* q, Euler3 *e)
+cdef quaternionToMatrix4(Matrix4 *m, Quaternion *q)
+cdef quaternionToEuler3(Euler3 *e, Quaternion *q)
```
Add `void` here. Put the `*` next to the parameter name, for instance `Quaternion* q` -> `Quaternion *q`.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -1,6 +1,11 @@
-from libc.math cimport atan2, sqrt, hypot, cos, sin
+cimport cython
+from libc.math cimport M_PI as PI, sqrt, hypot, abs, sin, cos, asin, acos, atan2, copysign
 from . matrix cimport normalizeMatrix_3x3_Part
 from . conversion cimport toPyEuler3
+from . vector cimport lengthVec3
+from . quaternion cimport setUnitQuaternion
+from mathutils import Vector, Matrix, Euler
+from .. math cimport quaternionNormalize_InPlace
```
Order the imports according to the line length. Short at the top, long at the button. Break long lines if necessary, see other files to see how this can be done.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +127,119 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+
+# https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef quaternionToEuler3(Euler3 *e, Quaternion *q):
+    quaternionNormalize_InPlace(q)
```
Lets assume all quaternions are normalized. So just remove all the necessary `quaternionNormalize_InPlace`.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +127,119 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+    cdef float yy = y * y
+    cdef float zz = z * z
+
+    cdef invs = 1 / (xx + yy + zz + ww)
```
Put `@cython.cdivision(True)` at the top of the function when the function have divisions or modulo. If the division can be a zero division, put an explicit check for zero.

OmarEmaraDev replied:
@kdsin23 This wasn't handled.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +127,119 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+
+    cdef invs = 1 / (xx + yy + zz + ww)
+
+    m.a11 = (xx - yy - zz + ww) * invs
+    m.a22 = (-xx + yy - zz + ww) * invs
+    m.a33 = (-xx - yy + zz + ww) * invs
+
+    m.a21 = 2.0 * (x * y + z * w) * invs
+    m.a12 = 2.0 * (x * y - z * w) * invs
+
+    m.a31 = 2.0 * (x * z - y * w) * invs
+    m.a13 = 2.0 * (x * z + y * w) * invs
+
+    m.a32 = 2.0 * (y * z + x * w) * invs
+    m.a23 = 2.0 * (y * z - x * w) * invs
```
Looks like you can group some expressions here as well. Can you try that?

OmarEmaraDev commented on animation_nodes/nodes/rotation/c_utils.pyx:
```diff
@@ -121,3 +127,65 @@ def randomQuaternionList(int seed, int amount):
+
+    for i in range(amount):
+        euler3ToQuaternion(qs.data + i, es.data + i)
+        quaternionNormalize_InPlace(qs.data + i)
```
Are the normalize calls here necessary? If they are, normalization should be done inside the `euler3ToQuaternion` function.

OmarEmaraDev reviewed (changes requested):
Use more consistent function naming, `quat` -> `quaternion`.

kdsin23 commented:
> quaternion

i am not sure we can name `quatFromAxisAngle` to `quaternionFromAxisAngle`, because  `quatFromAxisAngle` is the original function, if named, i afraid other place can not call this fuction any more.

OmarEmaraDev commented:
@kdsin23  `quatFromAxisAngle` is not used anywhere, so you can safely rename it.

kdsin23 commented:
> @kdsin23 `quatFromAxisAngle` is not used anywhere, so you can safely rename it.

OK, i have already done this above the code. what do i do next?

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pxd:
```diff
@@ -6,4 +6,9 @@ from . vector cimport Vector3
-cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, float cosAngle)
\ No newline at end of file
+cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, float cosAngle)
+cdef euler3ToQuaternion(Quaternion* q, Euler3 *e)
```
`void` here as well. `Quaternion* q` -> `Quaternion *q`.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -1,6 +1,14 @@
-from libc.math cimport atan2, sqrt, hypot, cos, sin
+cimport cython
+from libc.math cimport (
+    M_PI as PI, sqrt, hypot, abs, sin, cos, asin,
+    acos, atan2, copysign
+)
 from . matrix cimport normalizeMatrix_3x3_Part
 from . conversion cimport toPyEuler3
+from . vector cimport lengthVec3
+from . quaternion cimport setUnitQuaternion
+from mathutils import Vector, Matrix, Euler
+from .. math cimport quaternionNormalize_InPlace
```
Those are still not ordered. Longer lines at the button. And maybe put `from libc.math cimport` at the very end.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,127 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+##########################################################    
+        
+# https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef euler3ToQuaternion(Quaternion* q, Euler3 *e):
```
`Quaternion* q` -> `Quaternion *q`.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,127 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+# https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToAngle/index.htm
+@cython.cdivision(True)
+cdef void quaternionToAxisAngle(Vector3 *v, float *a, Quaternion *q):
+    quaternionNormalize_InPlace(q)
```
No need to normalize here.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,127 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+@cython.cdivision(True)
+cdef void quaternionToAxisAngle(Vector3 *v, float *a, Quaternion *q):
+    quaternionNormalize_InPlace(q)
+    cdef float k = sqrt(1 - q.w * q.w)
```
Can k be zero here?

kdsin23 replied:
how to fix? if `k = 0`, 

but i guess `k < 1` , because q is normalized.

kdsin23 replied:
you mean `cdef invs = 1 / (xx + yy + zz + ww)`changed to` cdef invs = xx + yy + zz + ww`?

OmarEmaraDev replied:
@kdsin23 If a zero case doesn't make sense, you should just set the output to a base case, like a zero or a unit vector.

kdsin23 replied:
`@cython.cdivision(True)
cdef void quaternionToAxisAngle(Vector3 *v, float *a, Quaternion *q):
    cdef float k = sqrt(1 - q.w * q.w)

    if k == 0:
        v.x = 1
        v.y = 0
        v.z = 0
    
    else:
        v.x = q.x / k
        v.y = q.y / k
        v.z = q.z / k
    
    a[0] = <float>(2 * acos(q.w))`

is this ok?

OmarEmaraDev replied:
@kdsin23 Yes, looks good.

kdsin23 replied:
Thank you for your so much guiding !

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,132 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+    cdef float yw = y * w
+    cdef float xw = x * w
+
+    cdef invs = xx + yy + zz + ww
+
+    m.a11 = (xx - yy - zz + ww) / invs
+    m.a22 = (-xx + yy - zz + ww) / invs
+    m.a33 = (-xx - yy + zz + ww) / invs
+
+    m.a21 = 2.0 * (xy + zw) / invs
+    m.a12 = 2.0 * (xy - zw) / invs
+
+    m.a31 = 2.0 * (xz - yw) / invs
+    m.a13 = 2.0 * (xz + yw) / invs
+
+    m.a32 = 2.0 * (yz + xw) / invs
+    m.a23 = 2.0 * (yz - xw) / invs
```
You are still not handling zero division here. Keep it as it was with `1 / (xx + yy + zz + ww)` but first check if `xx + yy + zz + ww` is zero, and if so, return a unit matrix.

```
cdef float dot = xx + yy + zz + ww
if dot == 0:
    setIdentityMatrix(q)
    return
cdef float invs = 1 / dot
```
Notice the `float` after cdef as well, which you seem to forgot.

OmarEmaraDev commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,132 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+        v.x = 1
+        v.y = 0
+        v.z = 0
+    
```
Remove this empty line.

kdsin23 commented on animation_nodes/math/rotation_conversion.pyx:
```diff
@@ -122,3 +130,132 @@ cdef void normalizedAxisCosAngleToMatrix(Matrix3_or_Matrix4* m, Vector3* axis, f
+##########################################################
+
+# https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+cdef void quaternionToEuler3(Euler3 *e, Quaternion *q):
```
i think matrix to quaternion is used original fuction `Matrix4x4List.toQuaternions()` to cacluate. so i don't know how to fix this issue.

the quaternion to euler fuction,i tested the different result,  i think the different results are the same angles. but i have no idea to adjust them the same. can you help me?

kdsin23 replied:
BTW, i can not go ahead patch-2, Why appear this:

To git@github.com:kdsin23/animation_nodes.git
 ! [rejected]        patch-2 -> patch-2 (non-fast-forward)

kdsin23 commented on CHANGELOG.md:
```diff
@@ -45,6 +45,7 @@
 - Increased default spline resolution for Spline nodes.
 - Remove redundant attribute setting in *Object Instancer* node.
 - Vectorized *Offset Splines* node.
+- Vectorized *Convert Rotations* node.
```
OK,no problem!

OmarEmaraDev replied:
@kdsin23 This line still needs to be removed.

OmarEmaraDev replied:
@kdsin23 No no, you removed the wrong line. I want to remove this one, the one in the unreleased section should be left as is.

OmarEmaraDev reviewed (changes requested):
You probably didn't merge correctly, because those files were never changed in master, but regardless, here is how to solve such conflicts.

When you get conflicts, git will add some markers in your file that follows the following pattern:

```
<<<<<<< HEAD
# Code you have in HEAD.
=======
# Code coming from the merge target.
>>>>>>> 1f02f68795d8a426250a239ea27999ddcacba2c2
```

To solve the conflict, you replace the lines starting from `<<<<<<< HEAD` to `>>>>>>> 1f02f687` with the correct lines, typically by choosing the code from `HEAD` or the code from the merge target or by choosing a combination of both.

So that's what you need to do. Go over the whole patch, when you see those markers, replace them with the correct code. I can also still see the CHANGELOG entry in the previous version, so that needs to be removed.

kdsin23 commented:
SO sorry for this request. why CHANGLOG is always conflicts? i am very confused about this.

OmarEmaraDev commented:
@kdsin23 CHANGELOG is always in conflict because all developers edit this file at the same time, so my changes my conflict with yours. I can't see the merge master commit here? Are you sure you merged master?

As I said, when there is a conflict, you update your repository, you merge master, and you fix conflicts. It seems you skipped the first two step.

kdsin23 commented:
> @kdsin23 CHANGELOG is always in conflict because all developers edit this file at the same time, so my changes my conflict with yours. I can't see the merge master commit here? Are you sure you merged master?
> 
> As I said, when there is a conflict, you update your repository, you merge master, and you fix conflicts. It seems you skipped the first two step.

yes, i  merge master  by website, not by git remote. you can see above change
`
Merge pull request #24 from kdsin23/master `

but i use `git merge origin/master` sometimes doesn't work, i don't know why

OmarEmaraDev commented:
Next time it doesn't work, show me the error. But don't merge master through a pull request, that is not what pull requests are for.

Nonetheless, the merge you did was 10 hours ago, master was updated 4 hours ago. So you still need to merge again. Try to do this now and I will guide you through it.

kdsin23 commented:
```
CHANGELOG.md: needs merge
error: you need to resolve your current index first
```

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git remote add upstream https://github.com/JacquesLucke/animation_nodes.git
fatal: remote upstream already exists.
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ git checkout masterCHANGELOG.md: needs merge
error: you need to resolve your current index first
ln@ln-pc:~/animation_nodes$ git fetch upstream
ln@ln-pc:~/animation_nodes$ git merge upstream/mastererror: merge is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.
ln@ln-pc:~/animation_nodes$ 

```

OmarEmaraDev commented:
@kdsin23 Looks like you already started a merge and didn't continue it. Reset it using `git reset --merge`, then attempt the merge again. Also, no need to execute `git remote add upstream` everytime, it is already created.

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git checkout patch-2
error: Your local changes to the following files would be overwritten by checkout:
	CHANGELOG.md
	animation_nodes/math/quaternion.pxd
	animation_nodes/math/quaternion.pyx
	animation_nodes/math/rotation_conversion.pxd
	animation_nodes/math/rotation_conversion.pyx
	animation_nodes/nodes/rotation/c_utils.pyx
	animation_nodes/nodes/rotation/convert_rotations.py
Please, commit your changes or stash them before you can switch branches.
```

OmarEmaraDev commented:
@kdsin23 Which branch are you in right now? `git status`

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   CHANGELOG.md
	modified:   animation_nodes/math/quaternion.pxd
	modified:   animation_nodes/math/quaternion.pyx
	modified:   animation_nodes/math/rotation_conversion.pxd
	modified:   animation_nodes/math/rotation_conversion.pyx
	modified:   animation_nodes/nodes/rotation/c_utils.pyx
	modified:   animation_nodes/nodes/rotation/convert_rotations.py

no changes added to commit (use "git add" and/or "git commit -a")

```

OmarEmaraDev commented:
@kdsin23 I see, you probably tried to merge patch-2 to master. Don't merge anything to master. Reset changes by `git restore .`. Then attempt to merge again.

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git restore 
git: 'restore' is not a git command. See 'git --help'.

Did you mean this?
```
	remote

OmarEmaraDev commented:
@kdsin23 You have an old version of git. Try `git checkout -- .`. Don't miss the dot at the end.

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git checkout -- ..
fatal: ..: '..' is outside repository
```

kdsin23 commented:
can i create a new branch and  pull a new request? i am very confused it,

OmarEmaraDev commented:
@kdsin23  Not two dots, a single dot.

kdsin23 commented:
```
lvan@lvan-pc:~/animation_nodes$ git checkout -- .
lvan@lvan-pc:~/animation_nodes$ 
lvan@lvan-pc:~/animation_nodes$ git checkout patch-2
Branch patch-2 set up to track remote branch patch-2 from origin.
Switched to a new branch 'patch-2'
lvan@lvan-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ git add .
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ git commit -m "fix errors."
On branch patch-2
Your branch is up-to-date with 'origin/patch-2'.
nothing to commit, working directory clean
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ 
ln@ln-pc:~/animation_nodes$ git push --set-upstream origin patch-2
Branch patch-2 set up to track remote branch patch-2 from origin.
Everything up-to-date
ln@ln-pc:~/animation_nodes$ 
```

kdsin23 commented:
what to do next?

OmarEmaraDev commented:
Okay, looks good. Now, try to update master and merge it into patch-2.

Are you on blender.chat by the way?

kdsin23 commented:
use the `git merge upstream/patch-2`?

kdsin23 commented:
no, i was in github chat with you

OmarEmaraDev commented:
```
git checkout master
git fetch upstream
git merge upstream/master
git checkout patch-2
git merge master
```

If all goes well, push. If there are conflicts, fix them, add changes, commit, then push.

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git merge upstream/patch-2
merge: upstream/patch-2 - not something we can merge
```

OmarEmaraDev commented:
Why are you trying to merge `upstream/patch-2`?

kdsin23 commented:
```
ln@ln-pc:~/animation_nodes$ git merge master
Auto-merging CHANGELOG.md
CONFLICT (content): Merge conflict in CHANGELOG.md
Automatic merge failed; fix conflicts and then commit the result.
```

kdsin23 commented:
> ```
> ln@ln-pc:~/animation_nodes$ git merge master
> Auto-merging CHANGELOG.md
> CONFLICT (content): Merge conflict in CHANGELOG.md
> Automatic merge failed; fix conflicts and then commit the result.
> ```

why this file was always wrong? how to correct the file?

OmarEmaraDev commented:
I already mentioned how we can fix conflicts, open the file, find the markers, and replace them. Then add the changes `git add CHANGELOG.md`, then commit `git commit -m "Merge master"`, then push `git push`.

OmarEmaraDev commented:
> @kdsin23 CHANGELOG is always in conflict because all developers edit this file at the same time, so my changes my conflict with yours.

kdsin23 commented:
OMG! it has been done well untill now! Thank you for your time!

kdsin23 commented:
is it OK now?

OmarEmaraDev reviewed (changes requested):
We want the line in the Unreleased section, not in the previous version section.

kdsin23 commented:
i think next time i will be more farmilar with this workflow, Thank you for your time!

OmarEmaraDev reviewed (approved):
Thanks!

I know the review processes was cumbersome for you. But you did great! You learned many new things and your next pull requests is going to go much smoother. Let me know if you need any further help with understanding git stuff.

kdsin23 commented:
@OmarEmaraDev  you are so nice and patient for me,  Thank you so much!
